### PR TITLE
Issue 457: llvmOPT=1_ASTCacheFix

### PR DESCRIPTION
**Pull request by ZGgof1999**
hmm, not sure if this makes the cache less useful, the same ASTs were needing different cfuncs, not sure if the error lies within the AST generation instead, but this gets things working. . . albeit slower than OPT=2

### CODE REVIEW

**Comment by geohot** [comment]
Huh, there should be a deeper fix than this. I suspect it's constant folding, add the constant to the str rep of the buffer like in the GPU.

**Comment by ZGgof1999** [comment]
right, right, I was wondering why the same key (AST) was sometimes needing an extra buf or more.. definitely screams something was left our or folded.
thanks for the pointer I'll check it out

**Comment by ZGgof1999** [comment]
yo I need to spend a little more time on this . . it doesn't seem like theres backing for the LLVM buffers, and thats where the GPU buffers are pulling the constants from no? A small explanation would go along way too i'm on the newer end of this stuff

**Comment by ZGgof1999** [comment]
@geohot I added a test and some notes in most recent push. . . added the constant stuff in the string but i think its a bigger issue :/ hopefully not

**Comment by geohot** [comment]
A lot of lines for this debugging, and I don't really understand what it is. Is there a short fix for this issue?

**Comment by ZGgof1999** [comment]
This is still a work in prog, and theres notes I added in 'external_test_llvm.py' in a new test... X.add(Y).add(Z) is okay, but X.add(X).add(X) computes wrong when MERGE_ELEMENTWISE_OPS is off in lazy.py. When reusing the same buffer the AST string in the cache gets confuzed.. I'm still looking for the fix, this last push was just a debugging session for me.

**Comment by geohot** [comment]
Marked PR as draft, lmk when ready to merge

**Comment by ZGgof1999** [comment]
adding constants wasn't the issue, but added it anyway, got messed up when RESHAPE ops were introduced so needed to track num bufs in cache key, added a test too

File: tinygrad/lazy.py [hunk 1]
```diff
@@ -372,4 +372,6 @@ def elementwise_op(op:Union[UnaryOps, BinaryOps], *srcs:LazyBuffer) -> LazyBuffe
     # remove the buffers from any (childless) BinaryOps that feed into this
     srcs = tuple(x.op if x.optype == BinaryOps and len(x.children) == 0 and x.realized is None else x for x in srcs)  # type: ignore
 
+  #print("")
```
Comment (geohot) [inline tinygrad/lazy.py:375]: Don't leave things like this

File: accel/llvm/ops_llvm.py [hunk 2]
```diff
@@ -170,9 +170,12 @@ def __init__(self, shape:Union[ShapeTracker, Tuple[int, ...]], hostbuf=None, for
     super().__init__(shape, hostbuf)
     # TODO: force alignment?
     self._buf = (ctypes.c_float * (prod(self.shape)))() if hostbuf is None else hostbuf._buf
+    #TODO: check with George if this method of constant tracking is okay . . . got from GPUBuffer structure
+    self._base_shape : Tuple[int, ...] = hostbuf._base_shape if hostbuf is not None else self.shape
```
Comment (geohot) [inline accel/llvm/ops_llvm.py:174]: Base shape here is fine, but is const even used? Does it break without it?

**Pull request closed without merging**